### PR TITLE
Determine grid scale factor automatically

### PR DESCRIPTION
**Pull request by dg0yt**
Revised variant of PR GH-1210, based on `dev`.

### CODE REVIEW

**Comment by dg0yt** [comment]
@pkturner This is revised series of your commits, now based on dev. I plan to do more changes before merging this. But at the moment I got two questions:

1. You chose a stereographic projection. I chose an orthographic projection for local projections (e.g. with GDAL/OGR data). Is there a good reason to prefer one over the other, in general or for a particular use case?

2. Do I understand correctly this PR still determines a single average scale factor, but we need separate X/Y scale factors for EPSG:3857?

BTW this change uncovered an existing error in `Georeferencing::load()`: It does not *skip* unknown elements. Thus the new `grid_compensation` element makes maps unreadable for older versions of Mapper. I have to fix this in some way before merging.

**Comment by pkturner** [comment]
@dg0yt It’s good to know you’re moving ahead with this.

1. I expect orthographic would work OK, since it is nearly the same as stereographic locally. I chose stereographic due to a love for conformal projections that I acquired as a math major in college.

I experimented with using the orthographic projection once. As I recall, it appeared to give a less accurate result when used to find the numeric derivative of the projection. However, either projection is basically accurate.

I suspect that the reason orthographic has been chosen over stereographic is that it may be faster to calculate.

2. That’s right. The added complexity is needed for any projection that distorts shapes ( is not conformal). 

Many of the projections that we use, e.g. UTM, are conformal or nearly so. In these cases it would be nice if Mapper could recognize the situation and use a simple scale factor. However, it would complicate some of the code, and I don’t know how to recognize the conformality in general.

**Comment by dg0yt** [comment]
I now [rewrote `GeoreferencingTest::testGridScaleFactor()`](https://github.com/OpenOrienteering/mapper/pull/1277/commits/f5b4c5593e7f40c4c696bebc020b366426ae0d88) to be data-driven, not comparing a single pair of projections, but the distance from map (paper) and scale against the geodesic distance (WGS84) from PROJ. IMO the test is much cleaner now.


Planned changes:

- I would prefer to not add an opaque variant of "grid compensation", but rather stick to scale factors (including y scale factor) and grivation if this is precise enough.
- When building against PROJ 5.1 or later (i.e. on most systems), we may simply get the scale factors from PROJ: https://proj.org/development/reference/functions.html#c.proj_factors

**Comment by pkturner** [comment]
> I would prefer to not add an opaque variant of "grid compensation", but rather stick to scale factors (including y scale factor) and grivation if this is precise enough. 

> When building against PROJ 5.1 or later (i.e. on most systems), we may simply get the scale factors from PROJ: https://proj.org/development/reference/functions.html#c.proj_factors

The projections we encounter developing orienteering maps often are either conformal (preserving shapes locally) like UTM, or have zero convergence (e.g. Simple Cylindrical or Web Mercator).  For conformal projections, convergence and scale factor are sufficent. For zero-convergence projections, horizontal and vertical scale factors are all that's needed.

However, when convergence and deformation occur together, as they do in most of the projections supported by PROJ https://proj.org/operations/projections/index.html, one more parameter is needed to specify the local effect of the projection. That's because when the scale factor is not uniform, the greatest scaling need not occur vertically or horizontally – it can occur in any direction.

From the description of PROJ's proj_factors on that web page, one would expect it to return all the partial derivatives relevant to convergence and scale factor. When starting out, I expected to use that in my code. But when I got to that part of the implementation, saw that the PJ_FACTORS data structure obviously does not cover the 4 degrees of freedom necessary for projections in general.

However those factors are determined, Mapper needs to combine them into a matrix in order to transform between map and projected coordinates. So it makes sense to represent them internally as a "grid compensation" matrix.

Mapper can handle this automatically for all projections. To limit the adjustments to the 3 degrees of freedom (convergence, x- and y- scale factors) would provide support for EPSG 3857, but miss the chance to support many others.

**Comment by pkturner** [comment]
> I now rewrote GeoreferencingTest::testGridScaleFactor() to be data-driven, not comparing a single pair of projections, but the distance from map (paper) and scale against the geodesic distance (WGS84) from PROJ.

That sounds good!

**Comment by dg0yt** [comment]
May I ask you to provide an example map with a relevant projection, location and shape?

**Comment by pkturner** [comment]
> May I ask you to provide an example map with a relevant projection, location and shape?

Of course. Are you asking for a real world example to demonstrate that people encounter deformations that cannot be handled with just x- and y- scale factors?

I've spent some time looking, and don't have an example orienteering map yet. Here in the US the surveys all use either Lambert conformal or UTM, so I don't expect my orienteering friends would be able to assist with an example. Neither do I have a set of maps to scan for examples, so I looked at [Wikipedia's list of national CRSes](https://en.wikipedia.org/wiki/List_of_National_Coordinate_Reference_Systems) to get an idea of where in the world it is likely to occur. That page shows only conformal systems for nations. So no leads there.

Would a contrived example be helpful, which makes ill-advised use of an orthographic projection, as an illustration of the type of deformation that's relevant?

**Comment by dg0yt** [comment]
A contrived example would be fine. It is to help teach developers what is going and why. Perhaps it can also be used in some future test, for example with an extra template in a different CRS.
(A contrived example is also used for the world file test.)

**Comment by pkturner** [comment]
Example files can be downloaded using [https://billygoat.org/Mapper/wintri-maps.zip](https://billygoat.org/Mapper/wintri-maps.zip). Except where noted, these were created using v0.8.3.

The example is based on an ordinary map that uses the local State Plane Coordinate System, EPSG:6575, which is a Lambert conformal projection. That's `junior.2019-05-24a.original.omap`.

`junior.2019-05-24a.naive-wintri.omap` was created from the original using the georeferencing dialog.

- custom proj spec: `+proj=wintri +ellps=WGS84 +units=m +no_defs`
- grid scale factor: `0.977167`

`junior.2019-05-24a.wintri.deformed.omap` was created as a new map, by first setting georeferencing with

- proj spec and grid scale factor as above
- reference point on map at 0, 0 mm
- reference point geographic 36.164085 N, -82.095455 W
- declination -7.08

Then the original map was imported.

`junior.2019-05-24a.wintri.omap` was created using a development build of Mapper, head of the grid-scale-factor-dev branch (commit 9838a986). It was created in the same way as the above "deformed" map, except that the grid scale factor was set automatically, as per the new default. This map is not readable by earlier versions of Mapper.

The example maps are georeferenced using the [Winkel tripel projection](https://en.wikipedia.org/wiki/Winkel_tripel_projection), which is designed to minimize distortion worldwide. As a consequence of tradeoffs, it deforms shapes everywhere except on the equator and prime meridian.

`junior.2019-05-24a.naive-wintri.omap` displays the incorrect relationship between map objects and geographic coordinates. When the GPS track `Track_2018-11-11 154238.gpx` is loaded, it appears in the wrong place.

`junior.2019-05-24a.wintri.deformed.omap` has all its objects in the correct geographic position, but they are skewed beyond any orienteering use by the Winkel projection. 

`junior.2019-05-24a.wintri.omap` works OK due to the grid-scale-factor-dev feature. The GPS track template loads correctly.

**Comment by pkturner** [comment]
@dg0yt 
> the new `grid_compensation` element makes maps unreadable for older versions of Mapper.

In the current feature, the user chooses “grid scale factor” “automatic” and that makes the map unreadable by older versions.

It may be unclear to some users why automatic calculation of the grid scale factor should create incompatibility. Actually, there's more to the incompatibility than the `grid_compensation` XML element causing a hiccup. If the projected coordinate system is deformed/anisotropic (not conformal) then older versions of Mapper are unable to calculate projected or geographic coordinates correctly.

What if the user were given three alternatives?
1. manual grid scale factor
2. automatic grid scale factor
3. support deformed/anisotropic projections (implies automatic grid scale factor)

with Mapper using and saving the `grid_compensation` matrix only if the 3rd is chosen.

It's tricky to draw a line between conformal and non-conformal projections so as to make the choice between 2 and 3 automatically. However, it would be relatively easy for the georeferencing dialog to warn if deformation/anisotropy has a substantial effect on the map, and suggest the 3rd alternative.

Here's an idea for how the file compatibility could work for users. Leave the version attribute at 7. Move the `grid_compensation` matrix to a top-level XML element, and generate it only if the map has support for deformed/anisotropic projections. Then a user of an old Mapper will encounter a warning if and only if the map has support for deformed/anisotropic projections.

**Comment by dg0yt** [comment]
The main issue with the `<grid_compensation>` XML element can be easily resolved by moving it below the `<project_crs>` element. The latter is an appropriate context, and it does not have the bug of the `<georeferencing>` element. And Mapper v0.8 could still degrade gracefully, using the saved (average) grid scale factor.

**Comment by dg0yt** [comment]
> wintri-maps.zip

Thanks for the nice example and explanations. But I must admit that this stirs my doubts about the grid compensation approach. But now I can express my concerns more clearly.

The question is: Do we want to add so much transformation in addition to the user-chosen CRS that it completely blurs the original projection? If the user explicitly selects a projection with heavy deformations, shouldn't this be respected, simply because this is the purpose of this selection?

For an o-map, it is reasonable to require the user to select a CRS based on more suitable projection. Worst case is EPSG 3857, with different x/y scale factors. (IIRC I added the scale factor field driven by issues reported for EPSG 3857.)

For other purposes, Mapper could draw maps in arbitrary projections. With some minimal code changes and a manually crafted GPX file drawing a bounding rectangle from 180° W to 180° E and from 50° N to 50° S, I was able to set an o-course around the equator in Winkel III projection:

world-map

Of course, the course length is not accurate. And if I try to display Lat/Lon coordinates, Mapper will live-lock if I exceed certain bounds. But in general, isn't it a more acceptable use case than o-mapping on Winkel III?

What we need for all type of maps is the reasonable precise placement of geospatial data (templates or import). The junior "deformed" map example shows that Mapper seems to work okay for GPX tracks. Of course we won't achieve satisfying results with rectangular raster images when just the corners are aligned to fit.

I do like the idea to warn the user about deformation/anisotropy having a substantial effect on the map.

**Comment by dg0yt** [comment]
Last push:
- Moved `<grid_compensation>` XML element to `<projected_crs>`.
- Test grid compensation effect on distances.
- Test "Winkel III" projection.

**Comment by pkturner** [comment]
> Do we want to add so much transformation in addition to the user-chosen CRS that it completely blurs the original projection? If the user explicitly selects a projection with heavy deformations, shouldn't this be respected, simply because this is the purpose of this selection?

No. The projection is chosen for georeferencing, so that the mapper can use templates, and so that a runner can overlay a track on the map for course review, etc. These purposes are secondary to orienteering itself. Orienteering relies on azimuths and scale. That's why Mapper adjusts for the projection's convergence and scale factor. The reason to adjust for deformation (heavy or slight) is the same.

> For an o-map, it is reasonable to require the user to select a CRS based on more suitable projection. Worst case is EPSG 3857, with different x/y scale factors. (IIRC I added the scale factor field driven by issues reported for EPSG 3857.)

Given that EPSG:3857 is the worst case deformation that is at issue, it may be acceptable to draw and print the map with that deformation, i.e. less than 1% scale error and less than 1° azimuth error. That much error would, however, cause trouble for mappers, if it affected the placement of templates.

> What we need for all type of maps is the reasonable precise placement of geospatial data (templates or import). The junior "deformed" map example shows that Mapper seems to work okay for GPX tracks. Of course we won't achieve satisfying results with rectangular raster images when just the corners are aligned to fit.

It appears to me that Qt is capable of skewing/deforming raster images, and that `TemplateTransform::fromQTransform` is unnecessarily throwing away some of that information. I have [a raster image](https://billygoat.org/Mapper/7_tile_slope.zip) that matches the example maps and illustrates the current misbehavior.

I'm not sure where that leaves us. Consider providing the basic “automatic grid scale factor” while removing compensation for deformation as in EPSG 3857 or Winkel III. IMO that change would not simplify the PR by much, nor would it be better for users.

**Comment by pkturner** [comment]
> I do like the idea to warn the user about deformation/anisotropy having a substantial effect on the map.

Here's a calculation of the anisotropic scale factors, the values that would determine whether to warn. 

    void Georeferencing::scaleFactorsOfCompensation(const QTransform &grid_compensation,
    double &greater_scale_factor, double &lesser_scale_factor)
    {
        // These are scale factors from true distance to grid distance.
        // They differ in case the projection has local deformation.
        const double m11 = grid_compensation.m11();
        const double m12 = grid_compensation.m12();
        const double m21 = grid_compensation.m21();
        const double m22 = grid_compensation.m22();
        const double sum = std::hypot(m11+m22, m12-m21);
        const double diff = std::hypot(m11-m22, m12+m21);
        greater_scale_factor = 0.5*(sum+diff);
        lesser_scale_factor = 0.5*(sum-diff);
     }

The warning could appear when the user clicks on the Georeferencing dialog's OK button.

**Comment by pkturner** [comment]
@dg0yt 
I am interested the possibility of making the code for this PR less "opaque" along the lines used for #1292. The idea is that the `grid_scale_factor` scalar and `grid_compensation` matrix would be replaced with `grid_scale_x`, `grid_scale_y`, and `grid_shear` members. Would you encourage this exploration?

There would be no loss of ability to handle EPSG 3857 or Winkel III.

**Comment by pkturner** [comment]
I have rebased from the OpenOrienteering:grid-scale-factor-dev branch, based on the current master. The update can be pulled from pkturner:grid-scale-factor-dev.

This is for the purpose of continuing work on the ideas discussed above.

**Comment by dg0yt** [comment]
At the moment it has conflicts in both `dev` and `master`. Let me see how to proceed...

**Comment by dg0yt** [comment]
> The update can be pulled from pkturner:grid-scale-factor-dev.

Ah. Got it.

File: src/gui/georeferencing_dialog.cpp [hunk 1]
```diff
@@ -385,6 +400,13 @@ void GeoreferencingDialog::declinationChanged()
 	setValueIfChanged(declination_edit, georef->getDeclination());
 }
 
+// slot
+void GeoreferencingDialog::gridScaleFactorChanged()
+{
+	const QSignalBlocker block(scale_factor_edit);
+	scale_factor_edit->setValue(georef->getGridScaleFactor());
+}
```
Comment (dg0yt) [inline src/gui/georeferencing_dialog.cpp]: If we should remove signal `Georeferencing::gridScaleFactorChanged()`, this code would go to `transformationChanged`.

**Comment by dg0yt** [comment]
Changed with the last pushes:
- Don't re-add obsolete native_file_format.cpp.
- Use QHBoxLayout instead QBoxLayout(left to right).

File: src/core/georeferencing.h [hunk 2]
```diff
@@ -565,14 +607,30 @@ friend QDebug operator<<(QDebug dbg, const Georeferencing& georef);
 	 */
 	void declinationChanged();
 	
+	/**
+	 * Indicates a change of the "grid scale factor", as presented in
+	 * the georeferencing dialog.
+	 * 
+	 * The "grid scale factor" has no direct influence on projection or transformation.
+	 * That's why there is an independent signal.
```
[the remark this answers is not in the document]
Comment (pkturner) [reply src/core/georeferencing.h]: This code mimics the `declinationChanged` signal that is used to keep the `declination_edit` box up-to-date. You're correct: the "grid scale factor" does in the `use_grid_compensation` case affect the transformation.

I'm not familiar with the tradeoff involved in having one more signal. If the `scale_factor_edit` box has its connection changed to the `transformationChanged` signal, then it appears that additional calls to `emit transformationChanged()` would be required.

File: src/core/georeferencing.cpp [hunk 3]
```diff
@@ -660,42 +699,78 @@ void Georeferencing::setScaleDenominator(int value)
...
-	if (grid_scale_factor != value)
+	double grid_scale_factor = roundScaleFactor(value);
+	if (grid_scale_factor != this->grid_scale_factor)
 	{
-		grid_scale_factor = value;
+		this->grid_scale_factor = grid_scale_factor;
+		if (!use_grid_compensation)
+		{
+			updateTransformation();
+			emit gridScaleFactorChanged();
```
[the remark this answers is not in the document]
Comment (pkturner) [reply src/core/georeferencing.cpp]: I hadn't thought about keeping the order of signals consistent.

File: src/gui/georeferencing_dialog.cpp (hunk 1 above)
Comment (pkturner) [reply src/gui/georeferencing_dialog.cpp]: Yes.

**Comment by pkturner** [comment]
Lately I've tended to join you, doubting the value of straightening out the deformation that occurs in a projection such as Winkel III. So I'm working on a variation in which the transformation uses just grivation and combined grid scale factor, rather than the full grid compensation matrix.

The Georeferencing dialog as it exists on this branch would work well with that less complex transformation. I'm also thinking about changing the `scale_factor_edit` from its current purpose of controlling the combined scale factor, and instead have it control the elevation factor.

The elevation factor is analogous to the declination -- both are suitable for user control.
The grid scale factor is analogous to the convergence -- calculated from the CRS and reference point.
The combined scale factor is analogous to the grivation -- used to build the transformation.

File: src/core/georeferencing.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dg0yt) [reply src/core/georeferencing.cpp]: That's a lesson that I learnt: You start to rely on the way an API works, and you get subtle errors from unexpected slight variations in behaviour. This is true in particular for Georeferencing, where we may have to update other items, e.g. various types templates,

In addition, I still need find a cut between the georeferencing data (copiable, no signals) and the observable georeferencing QObject (signals). At the moment, the class is a copiable QObject, and that is against best practice in Qt. This pending refactoring is another reason holding me back from adding more signals.

Last not least this particular signal is only used for the georeferencing dialog, and this is not on a hot path.

I will add a commit which removes the extra signal.

**Comment by dg0yt** [comment]
> Lately I've tended to join you, doubting the value of straightening out the deformation that occurs in a projection such as Winkel III. So I'm working on a variation in which the transformation uses just grivation and combined grid scale factor, rather than the full grid compensation matrix.

Thank you for letting me know. When I re-read the discussion yesterday I found that I am still not convinced our special-purpose mapping software would really need to support these types of CRS as primary georeferencing. (Handling "deformed" geospatial data in templates and import is a different story.)

**Comment by dg0yt** [comment]
PS: I still have an open browser tab for scale factor background:
https://www.surveyorpockettools.org/how-to-use/point-scale-factor

File: src/gui/georeferencing_dialog.cpp (hunk 1 above)
Comment (dg0yt) [reply src/gui/georeferencing_dialog.cpp]: Just found that the code is already there. Well, the (comined) grid scale factor is feature which came before this PR ;-)

**Comment by dg0yt** [comment]
One more proposed commit, https://github.com/OpenOrienteering/mapper/pull/1277/commits/f707cef08c52ffa3af6aed9716a40446502d0e0e: When using grid compensation, always update the grid scale factor on grid compensation changes.

**Comment by dg0yt** [comment]
>  I'm also thinking about changing the `scale_factor_edit` from its current purpose of controlling the combined scale factor, and instead have it control the elevation factor.
> 
> The elevation factor is analogous to the declination -- both are suitable for user control.
> The grid scale factor is analogous to the convergence -- calculated from the CRS and reference point.
> The combined scale factor is analogous to the grivation -- used to build the transformation.

I'm not sure how many mappers, novice or experienced, would know how to deal with elevation factor?

**Comment by pkturner** [comment]
> how many mappers, novice or experienced, would know how to deal with elevation factor?

Perhaps none, but I thought about them while developing this feature. I wished to eliminate the "Grid scale factor" widget from the Georeferencing dialog. However, I thought it had to be kept for the sake of possible mappers who care about the elevation factor.

**Comment by dg0yt** [comment]
>  I wished to eliminate the "Grid scale factor" widget from the Georeferencing dialog. However, I thought it had to be kept for the sake of possible mappers who care about the elevation factor.

What if we use a combo box to choose between grid compensation (= default), explicit factor (= legacy) and elevation factor (= advanced)? The factor input field would be disabled for the first option. We might even leave it blank if a single factor is not a good representation.

**Comment by dg0yt** [comment]
Abandoned in favor of #1508.

**Pull request closed without merging**